src/components/TablePagination/TablePagination.tsx

```diff
@@ -85,93 +45,145 @@ export type Props = PropsWithChildren<{
    * place the pagination component above or below the table?
    */
   position?: "above" | "below";
-}> &
+};
+
+export type ExternalControlProps = BasePaginationProps & {
+  /**
+   * Whether the component will be controlled via external state.
+   */
+  externallyControlled?: true;
+  /**
+   * the total number of items available within the data. This prop is only relevant
+   * and will be required if `externallyControlled` is set to `true`.
+   */
+  totalItems: number;
+  /**
+   * the current page that's showing. This prop is only relevant and will be required
+   * if `externallyControlled` is set to `true`.
+   */
+  currentPage: number;
+  /**
+   * size per page. This prop is only relevant and will be required if
+   * `externallyControlled` is set to `true`.
+   */
+  pageSize: number;
+  /**
+   * callback indicating a page change event to the parent component.
+   * This prop is only relevant and will be required if `externallyControlled` is set
+   * to `true`.
+   */
+  onPageChange: (page: number) => void;
+  /**
+   * callback indicating a page size change event to the parent component.
+   * This prop is only relevant and will be required if `externallyControlled` is set
+   * to `true`.
+   */
+  onPageSizeChange: (pageSize: number) => void;
+};
+
+export type InternalControlProps = BasePaginationProps & {
+  /**
+   * Whether the component will be controlled via external state.
+   */
+  externallyControlled?: false;
+};
+
+export type Props = PropsWithChildren<
+  BasePaginationProps & (ExternalControlProps | InternalControlProps)
```

Comment:
```suggestion
  ExternalControlProps | InternalControlProps
```
I think the `BasePaginationProps` are extended by both options.

Reply:
good catch, I must be in a rabbit hole at that point :sweat: